### PR TITLE
Fix source order sorting.

### LINKED ISSUE #2556: Types can end up in the wrong namespace
This is a regression from 83f729f44819459665832459cb86b68b4fc3739f

### Input C/C++ Header

I'm sorry this is not reduced as much as possible, but it's already significantly reduced. The problem does *not* happen using the output of --dump-preprocessed-input (it *does* require multiple separate headers, and even worse, it requires that some of them are included multiple times)

The complete set of files involved in this somehow reduced testcase is: https://drive.google.com/file/d/1vldgLGgcslyeUTlcHzMd4r0WVZNY2Wht/view?usp=sharing

### Bindgen Invocation

Assuming the `testcase.tar.gz` archive is extracted to `/tmp/testcase`, run the cli with
```
cargo run --release -p bindgen-cli -- /tmp/testcase/mozilla/GeckoBindings.h --allowlist-type nsSimpleContentList --allowlist-type nsSize --enable-cxx-namespaces --no-derive-default --generate types --ignore-methods --rust-target 1.68 -- -x c++ -std=c++17 -fno-sized-deallocation -fno-aligned-new -DTRACING=1 -DIMPL_LIBXUL -DMOZILLA_INTERNAL_API -DRUST_BINDGEN -I/tmp/testcase -I/tmp/testcase/nspr -include /tmp/testcase/mozilla-config.h -include /tmp/testcase/nsStyleStruct.h -include /tmp/testcase/LayoutConstants.h
```

(using clang 16, I don't know for sure whether the command works with older versions, I know for sure it doesn't work with clang 11, because I have that in a corner)

### Actual Results

The output contains the nsSize type in the root::mozilla namespace, although the type is defined in the root namespace.

There are many ways to make this disappear:
- removing `--allowlist-type nsSimpleContentList`
- in LayoutConstants.h, adding `constexpr nsSize dummy(0, 0);` above `namespace mozilla` (because yes, that's where the bogus namespace comes from)
- not including nsStyleStruct.h from the command line
- not including nsStyleStruct.h from mozilla/GeckoBindings.h
- using the preprocessed input
- creating a file that includes `mozilla-config.h`, `nsStyleStruct.h`, `LayoutConstants.h` and `mozilla/GeckoBindings.h`, and using that instead of passing those files on the command line.
- etc.

### PR DESCRIPTION
**Pull request by reitermarkus**
Fixes https://github.com/rust-lang/rust-bindgen/issues/2556.

Previously, headers included with `-include` were not added to the `includes` map since they don't have a source file. This is now fixed.

Also, some wrong match arms were removed/fixed.

### CODE REVIEW

**Comment by reitermarkus** [comment]
cc @glandium @emilio

File: bindgen-tests/tests/tests.rs [hunk 1]
```diff
@@ -468,9 +468,9 @@ fn test_mixed_header_and_header_contents() {
             env!("CARGO_MANIFEST_DIR"),
             "/tests/headers/func_ptr.h"
         ))
-        .header(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/headers/char.h"))
         .header_contents("test.h", "int bar(const char* a);")
         .header_contents("test2.h", "float bar2(const char* b);")
+        .header(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/headers/char.h"))
```
Comment (reitermarkus) [inline bindgen-tests/tests/tests.rs:473]: Note: This doesn't change anything, it simply reflects the order these are stored internally and eventually generated in.

File: bindgen/clang.rs [hunk 2]
```diff
@@ -555,36 +543,41 @@ impl Cursor {
...
-                if file2 == other_file2 {
-                    offset2.cmp(&other_offset2)
-                } else {
-                    cmp::Ordering::Equal
-                }
+            // If both items were included in the same file, compare the offset of their `#include` directives.
+            (Some((file2, offset2)), Some((other_file2, other_offset2)))
+                if file2 == other_file2 =>
+            {
+                offset2.cmp(&other_offset2)
```
Comment (emilio) [inline bindgen/clang.rs:577]: I don't think this is enough. You need to somehow deal with transitive includes don't you?

File: bindgen/clang.rs [hunk 3]
```diff
@@ -555,36 +543,41 @@ impl Cursor {
...
             }
-            (None, None) => cmp::Ordering::Equal,
+            // Otherwise, keep the original sorting.
+            _ => cmp::Ordering::Equal,
```
Comment (emilio) [inline bindgen/clang.rs:580]: I guess this gets the fallback, but is this ok? Seems this could end up with weird cases where the partial ord invariants are not held up and we end up with really weird sorting right?

File: bindgen/clang.rs (hunk 2 above)
Comment (reitermarkus) [reply bindgen/clang.rs:577]: Can you give an example for what you think might get sorted wrong? Will try to add a test for this then.

Note that when recursing, the recursive `visit_sorted` gets sorted in and of itself, so this may actually work without dealing with transitive includes explicitly.

File: bindgen/clang.rs (hunk 3 above)
Comment (reitermarkus) [reply bindgen/clang.rs]: Remove the fallback, all cases should be handled explicitly now.

File: bindgen/clang.rs (hunk 2 above)
Comment (emilio) [reply bindgen/clang.rs]: I don't have a concrete test-case because it would depend on the concrete sorting algorithm, but not dealing with recursive includes would mean that you break the `Ord` invariants, as in, you might have a, b, and c, such that `a < b`, `b < c`, but `a == c` or so, breaking https://doc.rust-lang.org/std/cmp/trait.Ord.html#corollaries, and potentially invariants of the sorting algorithm. Need to take a look at the new approach.

File: bindgen/clang.rs [hunk 4]
```diff
@@ -555,36 +543,154 @@ impl Cursor {
...
+                    //
+                    // a.h
+                    // ├── b.h
+                    // └── c.h
+                    //
+                    // When comparing items inside `b.h` and `c.h`, go up one level and
+                    // compare the include locations of `b.h` and `c.h` in `a.h` instead.
+                    include_location = ctx.included_file_location(file2);
+                    other_include_location =
+                        ctx.included_file_location(other_file2);
```
Comment (reitermarkus) [inline bindgen/clang.rs:671]: @emilio, I added another test which actually reaches this case.

**Comment by reitermarkus** [comment]
I refactored `SourceLocation` to be an `enum` and the sorting is now split between `SourceLocation` and `IncludeLocation`. In my opinion it is a bit easier to follow that way.

File: bindgen/clang.rs (hunk 2 above)
Comment (pvdrz) [reply bindgen/clang.rs]: I agree, we shouldn't rely on the fact that sort works for the testcases we have so far even though we don't have an actual ordering as it is not transitive. 

I'd be happy to fix this in an upcoming PR if that's ok with both of you.

File: bindgen/clang.rs (hunk 3 above)
Comment (pvdrz) [reply bindgen/clang.rs]: This might be a bit evil but what happens if you do the following:
```h
// a.h
#ifndef A_H
#define A_H

struct foo {};

#include "b.h"

struct baz {
    struct bar field;
};

#endif

```
```h
//b.h
#include "a.h"

struct bar {
  struct foo field;
};
``
Does this ordering guarantee that we don't loop forever because `a.h` is included in `b.h` which includes `a.h`?

Comment (reitermarkus) [reply bindgen/clang.rs]: @pvdrz, yes, we won't loop forever, but there is still an error here.

If `a.h` is the main header file and we compare `baz` with `bar`, we check the include location of `baz`, which is erroneously inside `b.h`, resulting in `bar` wrongly being sorted after `baz`.

The fix for this is easy though: Simply ignore include directives which include the main header.

Comment (pvdrz) [reply bindgen/clang.rs]: :+1: could you add this as a test?

**Comment by reitermarkus** [comment]
@pvdrz, I fixed the edge case you gave and added a test for it.

**Comment by reitermarkus** [comment]
I changed all file names from `String` to `Path` since we need to canonicalize them before comparing them, e.g. running your example with `cargo run -- a.h` would actually try to compare the strings `"a.h"` and `"./a.h"` before.

`ParseCallbacks::include_file` should probably also be changed to take a `&Path` instead of `&str`. Also its documentation

```rust
/// This will be called on every file inclusion, with the full path of the included file.
```

was actually wrong before (depending on whether you define “full path” as “absolute path”), since it would actually also pass relative paths.

**Review by emilio — COMMENTED** [review]
I still think this is quite complicated for the benefit it brings fwiw... :/

**Comment by reitermarkus** [comment]
@emilio, do you have an alternative? The status quo is simply  incorrect in some cases.

The only alternative I can see is to never generate constants for variable-like macros and only ever generate Rust macros instead. That way there is never a naming or ordering conflict.

Note that this is also needed for the same reason for function-like macros and functions in https://github.com/rust-lang/rust-bindgen/pull/2369.

**Comment by reitermarkus** [comment]
No, it only broke some cases due to the sorting being wrong/incomplete previously. Otherwise nothing should break, except for some auto-generated names having a different index.

**Comment by alex** [comment]
Confirmed that this works for one of my projects.

**Comment by reitermarkus** [comment]
Thanks for testing, @alex.

@emilio, @pvdrz, can you have another look at this?

**Comment by reitermarkus** [comment]
@pvdrz, @emilio, please have another look here, thank you.

**Comment by reitermarkus** [comment]
@pvdrz, @emilio, another ping here.

**Comment by reitermarkus** [comment]
@pvdrz, @emilio, can you have another look at this?

**Review by emilio — COMMENTED** [review]
I still think we should have some measurements of the overhead of the source order sorting, and turn it off by default if it's non-trivial.

Some of the refactoring like refactoring to use the clang::SourceLocation enum seems nice. Though that said, I'd rather keep the CXFile rather than reading the path and doing IO to absolutize it eagerly for most callers that don't need that...

**Comment by reitermarkus** [comment]
> turn it off by default

This is a prerequisite for supporting complex macros (https://github.com/rust-lang/rust-bindgen/pull/2369), so it can't be optional.

**Comment by reitermarkus** [comment]
@pvdrz, @emilio please have another look. 

> I'd rather keep the `CXFile` rather than reading the path and doing IO to absolutize it eagerly for most callers that don't need that...

I have re-added the `File` struct as `SourceFile` now to avoid eagerly absolutizing every path.

**Comment by reitermarkus** [comment]
@pvdrz, @emilio please have another look here.
